rule/sexpr/parser.go

```diff
@@ -23,3 +30,220 @@ func makeSymbolMap() *opCodeMap {
+			// case, because we should only ever parse a
+			// single tree of expressions (terminated by a
+			// right parenthesis), a single value or a
+			// single param.  This will become more cloudy
+			// if we start parsing comments.
+			return nil, newParserError(le, fmt.Errorf("unexpected end of file"))
+		case COMMENT:
+			// We ignore comments, for now
+			continue
+		case WHITESPACE:
+			// We just ignore white space, for now.
+			continue
+		case LPAREN:
+			if inOperator {
+				// This is a nested operator, so we'll
+				// want to treat it exactly like we
+				// did this one.  ... therefore, pop that left parenthesis back.
+				p.unscan()
+				// .... and recur!
+				expr, err = p.parseExpression()
+				if err != nil {
+					return nil, err
+				}
+				// Now we've got our nested
+				// expression, lets push it onto this
+				// one (and make sure it complies with
+				// the contract)
+				if err := opExpr.(rule.Operator).PushExpr(expr); err != nil {
+					// TODO: drastically improve this error message
+					return nil, newParserError(le, fmt.Errorf(
+						"Type mismatch in subexpression",
+					))
+				}
+				continue
+			}
+			// Left parenthesis must be followed by an operator
+			opExpr, err = p.parseOperator()
+			if err != nil {
+				return nil, err
+			}
+			inOperator = true
+
+		case BOOL:
+			expr = p.makeBoolValue(le)
+			if !inOperator {
+				// Great, we got a BoolValue, let's return that.
+				break Loop
+			}
+			// OK, let's push the BoolValue into our
+			// operator and see if that complies with the
+			// contract.
+			if err := opExpr.(rule.Operator).PushExpr(expr); err != nil {
+				return nil, newParserError(le, err)
+			}
+
+		case STRING:
+			expr = rule.StringValue(le.Literal)
+			if !inOperator {
+				// OK, we're done, break the loop
+				break Loop
+			}
+			// Lets see if our opExpr is really expecting a string
+			if err := opExpr.(rule.Operator).PushExpr(expr); err != nil {
+				return nil, newParserError(le, err)
+			}
+
+		case RPAREN:
+			if !inOperator {
+				// We don't have a matching LPAREN, so this is bad news.
+				return nil, newParserError(le, fmt.Errorf("Unexpected closing parenthesis"))
+			}
+			// We've close off the operator
+			inOperator = false
+			// .. lets finalise it, and see if that's compatible with the contract
+			if err := opExpr.(rule.Operator).Finalise(); err != nil {
+				return nil, newParserError(le, err)
+			}
+			// All is good, the output expression is this operater
+			expr = opExpr
+			break Loop
+		}
+
+	}
+	return expr, nil
+}
+
+// makeBoolValue extracts a BoolValue from a lexicalElement, or returns an error
+func (p *Parser) makeBoolValue(le *lexicalElement) rule.Expr {
+	if le.Literal == "true" {
+		return rule.BoolValue(true)
+	}
+	return rule.BoolValue(false)
+}
+
+// parseOperator attempts to convert the next symbol into an operator, if it fails and error is returned.
+func (p *Parser) parseOperator() (rule.Expr, error) {
```

Comment:
:tulip: Shouldn't this function return a rule.Operator?
This would allow us to avoid doing multiple type assertions above (lines 166, 190, 201, 213) and only do one l217.

Reply:
I did this, and it worked a treat ;-)